File: design-pattern-java/app/src/main/java/design/pattern/java/patterns/behavioral/iterator/FibonacciProducer.java
package design.pattern.java.patterns.behavioral.iterator;

public class FibonacciProducer implements Producer<Integer> {
  private Integer index;
  private Integer current;
  private Integer prev;

  public FibonacciProducer() {
    this.index = 0;
    this.current = 1;
    this.prev = 0;
  }

  public static Producer<Integer> create() {
    return new FibonacciProducer();
  }

  @Override
  public Boolean hasNext() {
    return this.index < 10;
  }

  @Override
  public Integer next() {
    if (!this.hasNext()) {
      throw new IndexOutOfBoundsException();
    }
    this.current += this.prev;
    this.prev = this.current - this.prev;
    this.index += 1;
    return this.current;
  }
}

File: design-pattern-jvm/app/src/main/kotlin/design/pattern/kotlin/patterns/behavioral/iterator/FibonacciProducer.kt
package design.pattern.kotlin.patterns.behavioral.iterator

import java.lang.IndexOutOfBoundsException

class FibonacciProducer : Producer<Int> {
    private var index = 0
    private var current = 1
    private var prev = 0
    override fun hasNext(): Boolean {
        return index < 10
    }

    override fun next(): Int {
        if (!hasNext()) {
            throw IndexOutOfBoundsException()
        }
        current += prev
        prev = current - prev
        index += 1
        return current
    }

    companion object {
        fun create(): Producer<Int> {
            return FibonacciProducer()
        }
    }
}